### ISSUE TITLE
[Bug]: "useArgs" hook updates wrong story when rendered in "Stories" block

### ISSUE THREAD

**Issue by brycenaddison** [storybookjs/storybook#28333]
### Describe the bug

When writing a story for a controlled component, one should expect the `useArgs` hook to update the args being passed to the story it is being called in. However, when used inside a "Stories" block, it updates the page's primary story instead of itself. This isn't an issue when using a "Canvas" block or any other block that selects a specific story. I'm sure this could be user error but I don't see why there would then be a discrepancy between the functionality between the "Stories" block and an individual "Story" block.

### Reproduction link

https://stackblitz.com/edit/github-n2sxrc?file=src%2Fstories%2FInput.tsx

### Reproduction steps

1. Go to the "Input" docs in Storybook.
2. Attempt to edit one of the stories under the "Stories" header.
3. Notice that value of top story is changing, instead of the story being edited.

### System

```bash
Storybook Environment Info:

  System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.20.3 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm <----- active
    pnpm: 8.15.6 - /usr/local/bin/pnpm
```

**Comment by brycenaddison** [comment]
For anyone having a similar issue, I was able to work around it by making a custom block to use instead of `Stories`.

```tsx
/** A block for showing a list of stories. */
export const MyStories = ({
  of,
  exclude = ['Primary'],
}): => {
  const resolvedOf = useOf(of ?? 'meta', ['meta']);
  const moduleExports = resolvedOf.csfFile.moduleExports;
  const exportsOrder = moduleExports.__namedExportsOrder;

  return (
    <>
      {exportsOrder
        .filter((name) => !exclude.includes(name))
        .map((name) => (
          // Title, description, and canvas blocks
          <MyStory key={name} of={moduleExports[name]} />
        ))}
    </>
  );
};

```

**Comment by asso1985** [comment]
+1

**Comment by TheSeydiCharyyev** [comment]
Cannot reproduce on `next` (10.5.0-alpha.0).

I added two stories using `useArgs()` to an autodocs file and inspected the `<Stories>` block on the docs page. Subscribed to `UPDATE_STORY_ARGS` on the manager channel and typed into the inputs — every event carried the correct `storyId`:

```
storyId: examples-button--controlled-b  updatedArgs: { label: 'B-initia' }
storyId: examples-button--controlled-b  updatedArgs: { label: 'B-initial ' }
storyId: examples-button--controlled-b  updatedArgs: { label: 'B-initialt' }
```

No leakage to the primary story. Same result with a button + click pattern. The underlying `useArgs`-returns-wrong-storyId bug appears resolved — no commit references this issue specifically, so the fix was likely incidental to preview-api context-isolation work between 8.3 and 10.x.

One related observation that may explain the original confusion: stories rendered inside `<Stories>` still pass `__forceInitialArgs={true}` (`code/addons/docs/src/blocks/blocks/Stories.tsx:68`). The args store updates the correct story, but the visible render is forced back to initial args on every update — so a controlled input in `<Stories>` looks unresponsive even though it's working. That's separate from this issue and might warrant a follow-up.

Safe to close unless someone can repro on the current `next`.

**Comment by Sidnioulz** [comment]
@TheSeydiCharyyev would you like to write a PR with an end-to-end test for this one? Happy to onboard you onto how to run them in VSCode (you gotta uncomment stuff in `code/playwright.config.ts` and build a React vite sandbox, then you can inspect every test or capture directly in the Storybook UI to write your test).

@shilman what do you think about `__forceInitialArgs`? If it causes re-renders to stories that attempt to call `useArgs`, then surely that's a UX bug?

**Comment by TheSeydiCharyyev** [comment]
Yes, I would like open a PR, but where should the test live? In a `code/e2e-tests/` alongside the existing ones, or somewhere specific to addon-docs?

**Comment by Sidnioulz** [comment]
In the `code/e2e-tests` folder sounds good to me, ideally alongside tests about args. Tests about framework decorators would fit well too, as this is where the hook is often called.

**Comment by TheSeydiCharyyev** [comment]
Started on this. There's no docspage template fixture using `useArgs` yet, so I'll add one — a controlled component (like the reporter's Input) with two stories calling `useArgs`, rendered in a Stories block.

A controlled `useArgs` render is renderer-specific, so I plan to scope the test to React sandboxes and skip the others, like the existing "source snippet should not change in stories block" test does. Does that fit, or would you prefer a renderer-agnostic fixture?

Since it no longer reproduces on `next`, the test will be a regression guard — asserting the arg update stays on the edited story, not the primary.
